File: src/platform/main_macos.zig
const core = @import("../modules/core.zig");

pub const m = struct {
    const macos = @import("../modules/macos.zig");
    pub usingnamespace macos;
    pub usingnamespace macos.libSystem;
    pub usingnamespace macos.libobjc;
    pub usingnamespace macos.fwFoundation;
    pub usingnamespace macos.fwAppKit;
};

pub const platform = struct {
    var app: *m.NSApplication = undefined;
    var running = true;
};

fn AsunderApplicationDelegate_applicationDidFinishLaunching(self: ?*anyopaque, sel: ?*anyopaque, notification: ?*anyopaque) callconv(.c) void {
    _ = .{ self, sel, notification };
    platform.app.@"stop:"(null);
}

fn AsunderWindowDelegate_windowWillClose(self: ?*anyopaque, sel: ?*anyopaque, notification: ?*anyopaque) callconv(.c) void {
    _ = .{ self, sel, notification };
    platform.running = false;
}

fn entry() void {
    inline for (@typeInfo(m.libobjc.classes).@"struct".decls) |decl| {
        @field(m.libobjc.classes, decl.name) = m.objc_getClass(decl.name).?;
    }
    inline for (@typeInfo(m.libobjc.selectors).@"struct".decls) |decl| {
        @field(m.libobjc.selectors, decl.name) = m.sel_getUid(decl.name).?;
    }

    const AsunderApplicationDelegate = m.objc_allocateClassPair(m.classes.NSObject, "AsunderApplicationDelegate", 0).?;
    _ = m.class_addMethod(AsunderApplicationDelegate, m.selectors.@"applicationDidFinishLaunching:", @ptrCast(&AsunderApplicationDelegate_applicationDidFinishLaunching), "v@:@");
    m.objc_registerClassPair(AsunderApplicationDelegate);

    const AsunderWindowDelegate = m.objc_allocateClassPair(m.classes.NSObject, "AsunderWindowDelegate", 0).?;
    _ = m.class_addProtocol(AsunderWindowDelegate, m.objc_getProtocol("NSWindowDelegate").?);
    _ = m.class_addMethod(AsunderWindowDelegate, m.selectors.@"windowWillClose:", @ptrCast(&AsunderWindowDelegate_windowWillClose), "v@:@");
    m.objc_registerClassPair(AsunderWindowDelegate);

    platform.app = m.NSApplication.sharedApplication().?;
    _ = platform.app.@"setActivationPolicy:"(.regular);

    const application_delegate = m.class_createInstance(AsunderApplicationDelegate, 0).?;
    platform.app.@"setDelegate:"(application_delegate);

    const window = m.NSWindow.alloc().?.@"initWithContentRect:styleMask:backing:defer:"(
        .{ .origin = .{ .x = 0, .y = 0 }, .size = .{ .w = 600, .h = 400 } },
        core.enumFlags(m.NSWindow.StyleMask, .{ .titled, .closable, .miniaturizable, .resizable }),
        .buffered,
        false,
    ).?;
    window.@"setTitle:"(m.NSString.alloc().?.@"initWithUTF8String:"("Asunder").?);

    const window_delegate = m.class_createInstance(AsunderWindowDelegate, 0).?;
    window.@"setDelegate:"(window_delegate);

    window.@"makeKeyAndOrderFront:"(null);

    platform.app.run();

    while (platform.running) {
        while (true) {
            const event = platform.app.@"nextEventMatchingMask:untilDate:inMode:dequeue:"(.any, null, m.NSDefaultRunLoopMode, @intFromBool(true));
            if (event == null) break;
            platform.app.@"sendEvent:"(event.?);
            event.?.release();
        }
    }
}

pub export fn _start() noreturn {
    entry();
    m.exit(0);
}

File: main_macos.c
// :todo replace objc messages with c calls.

#include "basic/basic.h"
#import <AppKit/AppKit.h> // :todo replace this with basic/macos.h

NSApplication* platform_app;
NSWindow* platform_window;
bool platform_running = true;

@interface AsunderApplicationDelegate : NSObject <NSApplicationDelegate>
@end

@implementation AsunderApplicationDelegate
-(void)applicationDidFinishLaunching:(NSNotification*)notification {
  [platform_app stop:nil];
}
@end

@interface AsunderWindowDelegate : NSObject <NSWindowDelegate>
@end

@implementation AsunderWindowDelegate
-(void)windowWillClose:(NSNotification*)notification {
  platform_running = false;
}
@end

noreturn_t start(void) {
  platform_app = [NSApplication sharedApplication];
  [platform_app setDelegate:[AsunderApplicationDelegate new]];
  [platform_app setActivationPolicy:NSApplicationActivationPolicyRegular];

  NSMenu* appbar = [NSMenu new];
  [platform_app setMainMenu:appbar];

  NSMenuItem* appmenuitem = [[NSMenuItem alloc] initWithTitle:@"Asunder" action:nil keyEquivalent:@""];
  [appbar addItem:appmenuitem];

  NSMenu* appmenu = [NSMenu new];
  [appmenuitem setSubmenu: appmenu];

  NSMenuItem* quitbutton = [[NSMenuItem alloc] initWithTitle:@"Quit Asunder" action:@selector(terminate:) keyEquivalent:@"q"];
  [appmenu addItem:quitbutton];

  platform_window = [[NSWindow alloc] initWithContentRect:(NSRect){(CGPoint){0.0, 0.0}, (CGSize){600.0, 400.0}} styleMask:NSWindowStyleMaskTitled|NSWindowStyleMaskResizable|NSWindowStyleMaskMiniaturizable|NSWindowStyleMaskClosable backing:NSBackingStoreBuffered defer:NO];
  [platform_window setDelegate:[AsunderWindowDelegate new]];
  [platform_window center];
  [platform_window setFrameAutosaveName:@"AsunderWindow"];
  [platform_window setTitle:@"Asunder"];
  [platform_window makeKeyAndOrderFront:nil];

  [platform_app run];

  for (;;) {
    for (;;) {
      NSEvent* event = [platform_app nextEventMatchingMask:NSEventMaskAny untilDate:nil inMode:NSDefaultRunLoopMode dequeue:YES];
      if (!event) break;
      [platform_app sendEvent:event];
      [event release];
    }
    if (!platform_running) break;
  }

  exit(0);
}
